### PR TITLE
Resolve Debug mode functions for the dynamic HAT package

### PR DESCRIPTION
**Pull request by lisaong**
* Resolve debug mode functions by adding a .o that includes the HAT file
* Switch to gcc rather than g++ so that we create C binaries

### CODE REVIEW

File: tools/hat_to_dynamic.py [hunk 1]
```diff
@@ -37,17 +37,26 @@ def get_platform():
...
     _, extension = os.path.splitext(input_hat_binary_path)
     if extension not in [".o", ".a"]:
         sys.exit(f"ERROR: Expected input library to have extension .o or .a, but received {input_hat_binary_path} instead")
 
+    # Create a C source file to resolve inline functions defined in the static HAT package
+    include_path = os.path.dirname(input_hat_binary_path)
+    inline_c_path = os.path.join(include_path, "inline.c")
+    inline_obj_path = os.path.join(include_path, "inline.obj")
+    with open(inline_c_path, "w") as f:
+        print("#include <{}>".format(os.path.basename(input_hat_path)), file=f)
```
Comment (kernhanda) [inline tools/hat_to_dynamic.py:52]: `f.write(f"#include <{os.path.basename(input_hat_path)}>")`?

File: tools/hat_to_dynamic.py [hunk 2]
```diff
@@ -37,17 +37,26 @@ def get_platform():
...
+    # compile it separately so that we can suppress the warnings about the missing terminating ' character
+    os.system(f'gcc -c -w -fPIC -o "{inline_obj_path}" -I"{include_path}" "{inline_c_path}"')
+
     # create new HAT binary
     prefix, _ = os.path.splitext(output_hat_path)
     output_hat_binary_path = prefix + ".so"
-    os.system(f'g++ -shared -fPIC -o "{output_hat_binary_path}" "{input_hat_binary_path}"')
+    os.system(f'gcc -shared -fPIC -o "{output_hat_binary_path}" "{inline_obj_path}" "{input_hat_binary_path}"')
```
Comment (kernhanda) [inline tools/hat_to_dynamic.py:59]: these os.system calls should probably be replaced by something a little more robust from the `subprocess` module. not critical, but it offers us better error reporting and output control

File: tools/hat_to_dynamic.py [hunk 3]
```diff
@@ -88,11 +97,12 @@ def windows_create_dynamic_package(input_hat_binary_path, output_hat_path, hat_d
...
-            with open("dllmain.cpp", "w") as f:
-                f.write("#include <windows.h>\n")
-                f.write("BOOL APIENTRY DllMain(HMODULE, DWORD, LPVOID) { return TRUE; }\n")
-        os.system("cl.exe /Fodllmain.obj /c dllmain.cpp")
+        with open("dllmain.cpp", "w") as f:
+            f.write("#include <windows.h>\n")
+            # Resolve inline functions defined in the static HAT package
+            f.write("#include <{}>\n".format(os.path.basename(input_hat_path)))
+            f.write("BOOL APIENTRY DllMain(HMODULE, DWORD, LPVOID) { return TRUE; }\n")
+        os.system('cl.exe /I"{}" /Fodllmain.obj /c dllmain.cpp'.format(os.path.dirname(input_hat_path)))
```
Comment (kernhanda) [inline tools/hat_to_dynamic.py:105]: we should use f-strings instead of the old .format(...) calls

**Review by kernhanda — APPROVED** [review]
lgtm

File: tools/hat_to_dynamic.py (hunk 2 above)
Comment (lisaong) [reply tools/hat_to_dynamic.py:59]: punting for future work.

**Pull request merged**